src/predictor/array_tree_layout.h

```diff
@@ -0,0 +1,250 @@
+/**
+ * Copyright 2021-2025, XGBoost Contributors
+ * \file array_tree_layout.cc
+ * \brief Implementation of array tree layout -- a powerfull inference optimization method.
+ */
+#ifndef XGBOOST_PREDICTOR_ARRAY_TREE_LAYOUT_H_
+#define XGBOOST_PREDICTOR_ARRAY_TREE_LAYOUT_H_
+
+#include <limits>
+#include <vector>
+
+namespace xgboost::predictor {
+
+/**
+ * @brief The class holds the array-based representation of the top levels of a single tree.
+ * 
+ * \tparam TreeType The type of the origianl tree (RegTree or MultiTargetTree)
+ *
+ * \tparam has_categorical if the tree has categorical features
+ *
+ * \tparam any_missing if the class is able to process missing values
+ * 
+ * \tparam kNumDeepLevels number of tree leveles being unrolled into array-based structure
+ */
+template <class TreeType, bool has_categorical, bool any_missing, int kNumDeepLevels>
+class ArrayTreeLayout {
+ private:
+  constexpr static size_t kNodesCount = (1u << kNumDeepLevels) - 1;
+
+  struct Empty {};
+  using DefaultLeftType =
+        typename std::conditional_t<any_missing,
+                                   std::array<uint8_t, kNodesCount>,
+                                   struct Empty>;
+  using IsCatType =
+        typename std::conditional_t<has_categorical,
+                                   std::array<uint8_t, kNodesCount>,
+                                   struct Empty>;
+  using CatSegmentType =
+        typename std::conditional_t<has_categorical,
+                                   std::array<common::Span<uint32_t const>, kNodesCount>,
+                                   struct Empty>;
+
+  DefaultLeftType default_left_;
+  IsCatType is_cat_;
+  CatSegmentType cat_segment_;
+
+  std::array<bst_feature_t, kNodesCount> split_index_;
+  std::array<float, kNodesCount> split_cond_;
+  std::array<bst_node_t, kNodesCount + 1> nidx_in_tree_;
+
+  inline static bool IsLeaf(const RegTree& tree, bst_feature_t nidx) {
+    return tree[nidx].IsLeaf();
```

Comment:
Please add this in order to prevent possible errors in compile time.
```suggestion
    static_assert(std::is_same_v<RegTree, T>); 
    return tree[nidx].IsLeaf();
```

Same applies to other similar cases.